Fail kubelet creation if swap enabled

mtaufen opened the pull request:
/cc @vishh @derekwaynecarr

**Release note**:

```
This adds an opt-in Kubelet flag (--fail-swap-on) that causes the Kubelet to fail to start if swap is enabled on the node. This is a temporary opt-in, and the Kubelet will by default fail with swap enabled starting in 1.6.0. The KubeletConfiguration equivalent to the flag is FailSwapOn.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31996)

<!-- Reviewable:end -->

mtaufen committed:
Fail kubelet creation if swap enabled

Provides an opt-in flag, --experimental-fail-swap-on (and corresponding
KubeletConfiguration value, ExperimentalFailSwapOn), which is false by default.

vishh commented on pkg/kubelet/kubelet.go:
```diff
@@ -350,6 +352,31 @@ func NewMainKubelet(kubeCfg *componentconfig.KubeletConfiguration, kubeDeps *Kub
 	if err != nil {
 		return nil, err
 	}
+	if len(thresholds) > 0 {
```
Don't check here. Move this logic to container_manager

derekwaynecarr replied:
I could see there or some NodeConfig type object that holds if swap is enabled that we can pass into EvictionManager and then have it error there...

mtaufen replied:
@vishh Any reason container manager would be more appropriate than the eviction manager? Eviction manager seems better to me as it already knows whether eviction is enabled, and wrt https://github.com/kubernetes/kubernetes/issues/31676#issuecomment-244496413 `eviction_enabled && swap_enabled` is the necessary condition for an error.

vishh replied:
Container manager is meant to validate node configuration wrt. to compute
resources. Eviction manager is meant to deal with evictions only. It should
not ideally care about system configuration.

On Tue, Sep 6, 2016 at 11:15 AM, Michael Taufen <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/kubernetes/kubernetes/pull/31996#discussion_r77689370>
> :
>
> > @@ -350,6 +352,31 @@ func NewMainKubelet(kubeCfg *componentconfig.KubeletConfiguration, kubeDeps *Kub
> >  	if err != nil {
> >  		return nil, err
> >  	}
> > +	if len(thresholds) > 0 {
>
> @vishh <https://github.com/vishh> Any reason container manager would be
> more appropriate than the eviction manager? Eviction manager seems better
> to me as it already knows whether eviction is enabled, and wrt #31676
> (comment)
> <https://github.com/kubernetes/kubernetes/issues/31676#issuecomment-244496413> eviction_enabled
> && swap_enabled is the necessary condition for an error.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31996/files/8c8e2fdb12920f5d56f1546ee676d87d008afd21#r77689370>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKPG5y-TJLnaahckBgRCXGOtA6x3Qks5qna3YgaJpZM4Jz-N7>
> .
>

mtaufen replied:
So there are two sorts of configuration that directly relate to this: Kubelet configuration (eviction) and node configuration (swap). The container manager clearly validates the latter, how much do we want it to validate the former? I'm ok with passing eviction information in if we decide to do so, but this could be a slippery slope that leads to passing more and more config fields into the container manager (remember what NewMainKubelet's argument list grew into not so long ago). Perhaps we should separate validation of node configuration and kubelet configuration?

mtaufen replied:
Granted, we also need a step that validates the combination of the two.

derekwaynecarr commented:
I would like this change as it would help reduce false errors in the
field.  Even though we recommend swap is off, mistakes grow like weeds.

On Friday, September 2, 2016, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *passed* for commit 8c8e2fd
> <https://github.com/kubernetes/kubernetes/commit/8c8e2fdb12920f5d56f1546ee676d87d008afd21>
> .
>
>    - Test Results
>    <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/31996/kubernetes-pull-build-test-e2e-gce/57173>
>    - Build Log
>    <http://pr-test.k8s.io/31996/kubernetes-pull-build-test-e2e-gce/57173/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/31996/kubernetes-pull-build-test-e2e-gce/57173/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//57173>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31996#issuecomment-244461778>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF8dbJa745GDO74CqUfdQvM0G6qlgSeGks5qmHOYgaJpZM4Jz-N7>
> .
>

mtaufen commented:
@vishh I moved the validation to the container manager, but I am still not sure that this is the cleanest place to do this validation. I think this is fine for now, but we need to have a discussion around where different configuration sources should be validated in the Kubelet. Maybe we do this as part of our next refactoring discussion.

derekwaynecarr commented on cmd/kubelet/app/server.go:
```diff
@@ -405,15 +406,24 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
+		if err != nil {
+			return err
+		}
+		evictionEnabled := len(thresholds) > 0
```
if i setup eviction on disk and not memory, then this is wrong.

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -164,7 +166,40 @@ func validateSystemRequirements(mountUtil mount.Interface) (features, error) {
+	//                interaction of KubeletConfiguration and node configuration. I'm not sure whether this should happen in
+	//                the container manager or not. It feels kind of weird to leak the KubeletConfiguration into this module.
+
+	if evictionEnabled {
```
this should eviction based on memory only.

mtaufen commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[cmd/kubelet/app/server.go, line 413 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31996#-KRAdaXRtT5sm_U_M9Bq-r2-413:-KRFvNek6pyo4Vtpj1hK:b-68dccb) ([raw file](https://github.com/kubernetes/kubernetes/blob/033c9e52311e5d091dddcfea5be37e35e8582c86/cmd/kubelet/app/server.go#L413)):*
<details><summary><i>Previously, derekwaynecarr (Derek Carr) wrote…</i></summary>
> if i setup eviction on disk and not memory, then this is wrong.
</details>
Updated to look for thresholds that use `SignalMemoryAvailable`.

---

*[pkg/kubelet/cm/container_manager_linux.go, line 179 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31996#-KRAdaXRtT5sm_U_M9Br-r2-179:-KRFvLrtB7NfYUoR_CoY:b-bw8e14) ([raw file](https://github.com/kubernetes/kubernetes/blob/033c9e52311e5d091dddcfea5be37e35e8582c86/pkg/kubelet/cm/container_manager_linux.go#L179)):*
<details><summary><i>Previously, derekwaynecarr (Derek Carr) wrote…</i></summary>
> this should eviction based on memory only.
</details>
Updated.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31996)*
<!-- Sent from Reviewable.io -->

mtaufen commented:
@derekwaynecarr @vishh I finally had time to rebase this. PTAL.

mtaufen commented:
Also @jdef this will necessitate a change in mesos, because `NewContainerManager` gets a new arg (`memoryEvictionEnabled`) in this PR. Change needed here: https://github.com/kubernetes-incubator/kube-mesos-framework/blob/fd9531a578c6c46daaa1e01621efa7e8f4dfe501/pkg/executor/service/service.go#L223

jdef commented:
thanks for the heads up.
/cc @k82cn

k82cn commented:
@mtaufen , @jdef , thanks for your heads up; I'll handle it in https://github.com/kubernetes-incubator/kube-mesos-framework/issues/33

vishh commented:
I'd fail kubelet if swap is enabled. We can provide an option to opt-out of this behavior. We do not support swap in many fronts.

mtaufen commented:
@vishh I am super in favor of that. I suppose we'll have to turn on the opt-out by default for at least 1.5 so we don't break anyone without a release-cycle's worth of warning, correct?

vishh commented:
Yeah. Add it to the list of kernel validation we already perform. If
possible, we can disable swap automatically.

On Tue, Oct 11, 2016 at 11:30 AM, Michael Taufen <notifications@github.com>
wrote:

> @vishh <https://github.com/vishh> I am super in favor of that. I suppose
> we'll have to turn on the opt-out by default for at least 1.5 so we don't
> break anyone without a release-cycle's worth of warning, correct?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31996#issuecomment-253003303>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKMU_OxsFEYwq9kHXW-4-mu6Wj-q_ks5qy9W8gaJpZM4Jz-N7>
> .
>

derekwaynecarr commented:
@vishh -- if we fail kubelet with swap enabled, we need to opt in for that behavior.

mtaufen commented:
@derekwaynecarr @vishh I changed this to provide an opt-in to fail with swap enabled instead of using eviction thresholds as the cue. 

@jdef @k82cn this changes the semantics of the necessary change to mesos, but the location should be the same.

mtaufen commented:
IMO we should target failing with swap enabled as a default behavior for 1.6, what do you guys think?

k82cn commented:
@mtaufen , thanks for your reminder; I think we can merge this PR firstly,  kube upstream was pinned in `kube-mesos-framework`, so it'll not break kube-mesos-frameworks' build.

I'm working kube-mesos's e2e test; when it's ready, I'll update kube-mesos's code accordingly.

mtaufen commented:
Ok sounds good. Thanks Klaus!

vishh commented:
Irrespective of the release, this change will surprise users. IMO, the
earlier we do it, the better it is. If we have it opt-in, then users will
have to react to swap+eviction issues. If we make it opt-out, they will
have to either fix their machines or update k'let configuration
acknowledging that evictions won't work.

On Thu, Oct 13, 2016 at 9:39 AM, k8s-ci-robot <notifications@github.com>
wrote:

> Jenkins verification *failed*
> <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/31996/kubernetes-pull-verify-all/18229/>
> for commit ab02ef8
> <https://github.com/kubernetes/kubernetes/commit/ab02ef8230d9a22a777fa5305979354fa2aa70b0>.
> Full PR test history <http://pr-test.k8s.io/31996>.
>
> The magic incantation to run this job again is @k8s-bot verify test this.
> Please help us cut down flakes by linking to an open flake issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+label:kind/flake+is:open>
> when you hit one in your PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31996#issuecomment-253568171>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKMZQBLab0g7uAg8eAlOSHUkgycPqks5qzl7ZgaJpZM4Jz-N7>
> .
>

vishh commented:
@derekwaynecarr @mtaufen For v1.5, let's fail kubelet only if swap is enabled and eviction is requested. Let's print a warning if swap is enabled in general and state that from v1.6 kubelet will fail is swap is enabled. That way there will hopefully be less surprises to users.

mtaufen commented:
One thing that worries me wrt surprises is that eviction is on by default, so failing on `swap && eviction` is actually pretty close to just failing on `swap`.

For this reason it might be better to maintain the `--fail-swap-on` approach for v1.5, with the addition that when `swap` is detected, we always print a warning that in v1.6 this node configuration will be rejected by the Kubelet.

vishh commented:
On Mon, Oct 31, 2016 at 3:04 PM, Michael Taufen <notifications@github.com>
wrote:

> One thing that worries me wrt surprises is that eviction is on by default,
> so failing on swap && eviction is actually pretty close to just failing
> on swap.
>
Ahh that's a good point. Agreed. This change will affect users no matter
what we do. May be this change can be phrased as "making kubelet fail
loudly" rather than "changing kubelet behavior". WDYT?

> For this reason it might be better to maintain the --fail-swap-on
> approach for v1.5, with the addition that when swap is detected, we
> always print a warning that in v1.6 this node configuration will be
> rejected by the Kubelet.
>
Ack!

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31996#issuecomment-257434336>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKL-bi7O4pRtmqt5B_7e5LvgYkLgGks5q5mXVgaJpZM4Jz-N7>
> .
>

mtaufen commented:
Can we phrase it as an opt-in enforcement of existing node configuration rules, which will be enforce by default in v1.6? 

I want to give some urgency to opting-in before v1.6 is released, rather than give the impression that it is nothing more than a helpful, optional error.

vishh commented:
As you said earlier, there is no way for us to make this opt-in since
memory eviction is opt-out. The only way users can opt-out of this failure
is by disabling memory evictions. Given that scenario, do we even need a
special opt-out flag for avoiding swap checks?

On Tue, Nov 1, 2016 at 1:31 PM, Michael Taufen <notifications@github.com>
wrote:

> Can we phrase it as an opt-in enforcement of existing node configuration
> rules, which will be enforce by default in v1.6?
>
> I want to give some urgency to opting-in before v1.6 is released, rather
> than give the impression that it is nothing more than a helpful, optional
> error.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31996#issuecomment-257686039>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKEbZpRCyT5dUJTTT16qKiHAt5KR9ks5q56GfgaJpZM4Jz-N7>
> .
>

mtaufen commented:
@vishh and I had an offline discussion, this is the plan we agreed on:

1. Always print a warning on Kubelet startup if swap is enabled.
2. Provide the `--experimental-fail-swap-on` opt-in option in 1.5 for people to canary this behavior before upgrading 1.5 to 1.6.
3. Cherry-pick (1) and (2) into 1.4 so people running 1.4 are notified of the policy and can canary this behavior before upgrading 1.4 to 1.6.
4. Document and warn people of this soon-to-be-default change in the release notes, etc. for 1.5 and the next 1.4 patch release. 
5. Remove the flag in 1.7 or later. 

IMO we probably shouldn't remove the flag until failing if swap is enabled is the standard behavior in all supported upgrade source versions (1.6, 1.7 to 1.8 will be the first time this occurs). We don't want someone to have to turn off enforcement prior to an upgrade to have the same node validation behavior on both sides. The Kubelet rejects unknown flags, and thus *might* be unable to start after an upgrade if the flag is no longer supported, depending on how a user's flags are provisioned.

mtaufen commented:
@vishh @derekwaynecarr PTAL
